ruma-federation-api/src/transactions/edu.rs

```diff
@@ -0,0 +1,432 @@
+    pub data: Receipt,
+
+    /// The extremity event ID the user has read up to.
+    pub event_ids: Vec<EventId>,
+}
+
+impl ReceiptData {
+    /// Creates a new `ReceiptData`.
+    pub fn new(data: Receipt, event_ids: Vec<EventId>) -> Self {
+        Self { data, event_ids }
+    }
+}
+
+/// The content for "m.typing" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct TypingContent {
+    /// The room where the user's typing status has been updated.
+    pub room_id: RoomId,
+
+    /// The user ID that has had their typing status changed.
+    pub user_id: UserId,
+
+    /// Whether the user is typing in the room or not.
+    pub typing: bool,
+}
+
+impl TypingContent {
+    /// Creates a new `TypingContent`.
+    pub fn new(room_id: RoomId, user_id: UserId, typing: bool) -> Self {
+        Self { room_id, user_id, typing }
+    }
+}
+
+/// The description of the direct-to- device message.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct DeviceListUpdateContent {
+    /// The user ID who owns the device.
+    pub user_id: UserId,
+
+    /// The ID of the device whose details are changing.
+    pub device_id: DeviceIdBox,
+
+    /// The public human-readable name of this device. Will be absent if the device has no name.
+    pub device_display_name: String,
+
+    /// An ID sent by the server for this update, unique for a given user_id.
+    pub stream_id: UInt,
+
+    /// The stream_ids of any prior m.device_list_update EDUs sent for this user
+    /// which have not been referred to already in an EDU's prev_id field.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub prev_id: Vec<UInt>,
+
+    /// True if the server is announcing that this device has been deleted.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub deleted: Option<bool>,
+
+    /// The updated identity keys (if any) for this device.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub keys: Option<DeviceKeys>,
+}
+
+impl DeviceListUpdateContent {
+    /// Creates a new `DeviceListUpdateContent` with default `prev_id`, `deleted`
+    /// and `keys` fields.
+    pub fn new(user_id: UserId, device_id: DeviceIdBox, name: String, stream_id: UInt) -> Self {
+        Self {
+            user_id,
+            device_id,
+            device_display_name: name,
+            stream_id,
+            prev_id: vec![],
+            deleted: None,
+            keys: None,
+        }
+    }
+}
+
+/// The description of the direct-to- device message.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct DirectDeviceContent {
+    /// The user ID of the sender.
+    pub sender: UserId,
+
+    /// Event type for the message.
+    #[serde(rename = "type")]
+    pub ev_type: String,
```

Comment:
Why not `ruma_events::EventType`?

Reply:
I wasn't sure if it was the same thing but it appears to be. I fixed it.